cpu/esp32: move ESP32_SDK_DIR definition here

maribu opened the pull request:
### Contribution description

The definition in `pkg/esp32_sdk/Makefile.include` was evaluated by `make` after the include paths were already set, resulting in `ESP32_SDK_DIR` being empty in

    INCLUDES += -I$(ESP32_SDK_DIR)/components
    [...]

This in turn resulted in

    cc1: error: /components: No such file or directory [-Werror=missing-include-dirs]
    [...]


### Testing procedure

#### In `master`

```
$ make BOARD=esp32-mh-et-live-minikit -C examples/hello-world
make: Entering directory '/home/maribu/Repos/software/RIOT/examples/hello-world'
Building application "hello-world" for "esp32-mh-et-live-minikit" with MCU "esp32".

"make" -C /home/maribu/Repos/software/RIOT/pkg/esp32_sdk/ 
"make" -C /home/maribu/Repos/software/RIOT/boards/common/init
"make" -C /home/maribu/Repos/software/RIOT/boards/esp32-mh-et-live-minikit
"make" -C /home/maribu/Repos/software/RIOT/boards/common/esp32
"make" -C /home/maribu/Repos/software/RIOT/core
"make" -C /home/maribu/Repos/software/RIOT/core/lib
"make" -C /home/maribu/Repos/software/RIOT/cpu/esp32
cc1: error: /components: No such file or directory [-Werror=missing-include-dirs]
cc1: error: /components/bootloader_support/include: No such file or directory [-Werror=missing-include-dirs]
cc1: error: /components/driver/include: No such file or directory [-Werror=missing-include-dirs]
cc1: error: /components/esp_common/include: No such file or directory [-Werror=missing-include-dirs]
cc1: error: /components/esp_hw_support/include: No such file or directory [-Werror=missing-include-dirs]
cc1: error: /components/esp_hw_support/include/soc: No such file or directory [-Werror=missing-include-dirs]
cc1: error: /components/esp_rom/esp32: No such file or directory [-Werror=missing-include-dirs]
cc1: error: /components/esp_rom/include: No such file or directory [-Werror=missing-include-dirs]
cc1: error: /components/esp_rom/include/esp32: No such file or directory [-Werror=missing-include-dirs]
cc1: error: /components/esp_system/include: No such file or directory [-Werror=missing-include-dirs]
cc1: error: /components/esp_system/port/include: No such file or directory [-Werror=missing-include-dirs]
cc1: error: /components/esp_timer/include: No such file or directory [-Werror=missing-include-dirs]
cc1: error: /components/hal/esp32/include: No such file or directory [-Werror=missing-include-dirs]
cc1: error: /components/hal/include: No such file or directory [-Werror=missing-include-dirs]
cc1: error: /components/hal/platform_port/include: No such file or directory [-Werror=missing-include-dirs]
cc1: error: /components/heap/include: No such file or directory [-Werror=missing-include-dirs]
cc1: error: /components/log/include: No such file or directory [-Werror=missing-include-dirs]
cc1: error: /components/newlib/platform_include: No such file or directory [-Werror=missing-include-dirs]
cc1: error: /components/soc/include: No such file or directory [-Werror=missing-include-dirs]
cc1: error: /components/soc/esp32/include: No such file or directory [-Werror=missing-include-dirs]
cc1: error: /components/xtensa/include: No such file or directory [-Werror=missing-include-dirs]
cc1: error: /components/xtensa/esp32/include: No such file or directory [-Werror=missing-include-dirs]
In file included from /home/maribu/Repos/software/RIOT/cpu/esp32/include/periph_cpu.h:24,
                 from /home/maribu/Repos/software/RIOT/boards/common/esp32/include/periph_conf_common.h:27,
                 from /home/maribu/Repos/software/RIOT/boards/esp32-mh-et-live-minikit/include/periph_conf.h:166,
                 from /home/maribu/Repos/software/RIOT/boards/common/esp32/include/board_common.h:32,
                 from /home/maribu/Repos/software/RIOT/boards/esp32-mh-et-live-minikit/include/board.h:48,
                 from /home/maribu/Repos/software/RIOT/cpu/esp32/startup.c:28:
/home/maribu/Repos/software/RIOT/cpu/esp32/include/sdkconfig.h:42:10: fatal error: soc/soc_caps.h: No such file or directory
   42 | #include "soc/soc_caps.h"
      |          ^~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
compilation terminated.
make[2]: *** [/home/maribu/Repos/software/RIOT/Makefile.base:146: /home/maribu/Repos/software/RIOT/examples/hello-world/bin/esp32-mh-et-live-minikit/cpu/startup.o] Error 1
make[1]: *** [/home/maribu/Repos/software/RIOT/Makefile.base:31: ALL--/home/maribu/Repos/software/RIOT/cpu/esp32] Error 2
make: *** [/home/maribu/Repos/software/RIOT/examples/hello-world/../../Makefile.include:738: application_hello-world.module] Error 2
make: Leaving directory '/home/maribu/Repos/software/RIOT/examples/hello-world'
```

#### This PR:

```
$ make BOARD=esp32-mh-et-live-minikit -C examples/hello-world
make: Entering directory '/home/maribu/Repos/software/RIOT/examples/hello-world'
Building application "hello-world" for "esp32-mh-et-live-minikit" with MCU "esp32".

"make" -C /home/maribu/Repos/software/RIOT/pkg/esp32_sdk/ 
"make" -C /home/maribu/Repos/software/RIOT/boards/common/init
"make" -C /home/maribu/Repos/software/RIOT/boards/esp32-mh-et-live-minikit
"make" -C /home/maribu/Repos/software/RIOT/boards/common/esp32
"make" -C /home/maribu/Repos/software/RIOT/core
"make" -C /home/maribu/Repos/software/RIOT/core/lib
"make" -C /home/maribu/Repos/software/RIOT/cpu/esp32
/home/maribu/Repos/software/RIOT/cpu/esp32/startup.c: In function 'system_init':
/home/maribu/Repos/software/RIOT/cpu/esp32/startup.c:243:31: error: implicit conversion from 'enum <anonymous>' to 'esp_log_level_t' [-Werror=enum-conversion]
  243 |     esp_log_level_set("wifi", LOG_DEBUG);
      |                               ^~~~~~~~~
/home/maribu/Repos/software/RIOT/cpu/esp32/startup.c:244:31: error: implicit conversion from 'enum <anonymous>' to 'esp_log_level_t' [-Werror=enum-conversion]
  244 |     esp_log_level_set("gpio", LOG_DEBUG);
      |                               ^~~~~~~~~
cc1: all warnings being treated as errors
make[2]: *** [/home/maribu/Repos/software/RIOT/Makefile.base:146: /home/maribu/Repos/software/RIOT/examples/hello-world/bin/esp32-mh-et-live-minikit/cpu/startup.o] Error 1
make[1]: *** [/home/maribu/Repos/software/RIOT/Makefile.base:31: ALL--/home/maribu/Repos/software/RIOT/cpu/esp32] Error 2
make: *** [/home/maribu/Repos/software/RIOT/examples/hello-world/../../Makefile.include:738: application_hello-world.module] Error 2
make: Leaving directory '/home/maribu/Repos/software/RIOT/examples/hello-world'
```

### Issues/PRs references

None

maribu commented:
@aabadie: I also dropped the `export` as requested.

gschorcht commented:
> The definition in `pkg/esp32_sdk/Makefile.include` was evaluated by `make` after the include paths were already set, resulting in `ESP32_SDK_DIR` being empty in

Hm, I'm a bit confused why this change is necessary, it has been working all the time as it was. Why is there now the problem that `ESP32_SDK_DIR` is not defined? I can't reproduce the problem in master.

maribu commented:
It does work also fine for me with `BUILD_IN_DOCKER=1`.

The issue is that if it works `INCLUDES` is recursively expended. This can be tested with:

```diff
diff --git a/cpu/esp32/Makefile.include b/cpu/esp32/Makefile.include
index d1123dfd75..010eb72d77 100644
--- a/cpu/esp32/Makefile.include
+++ b/cpu/esp32/Makefile.include
@@ -48,6 +48,8 @@ PSEUDOMODULES += esp_spi_ram
 PSEUDOMODULES += esp_spi_oct
 PSEUDOMODULES += esp_wifi_enterprise
 
+$(info INCLUDES (cpu): $(origin INCLUDES), $(flavor INCLUDES))
+
 INCLUDES += -I$(RIOTCPU)/$(CPU)/esp-idf/include
 INCLUDES += -I$(RIOTCPU)/$(CPU)/esp-idf/include/log
 INCLUDES += -I$(RIOTCPU)/$(CPU)/vendor/include
diff --git a/pkg/esp32_sdk/Makefile.include b/pkg/esp32_sdk/Makefile.include
index d028aa7943..2c017c0f72 100644
--- a/pkg/esp32_sdk/Makefile.include
+++ b/pkg/esp32_sdk/Makefile.include
@@ -1,3 +1,6 @@
 export ESP32_SDK_DIR ?= $(PKGDIRBASE)/esp32_sdk
 
+$(info INCLUDES (pkg): $(origin INCLUDES), $(flavor INCLUDES))
+$(info INCLUDES = $(INCLUDES))
+
 PSEUDOMODULES += esp32_sdk
```

When it fails `INCLUDES` has been converted to simple expension (e.g. by `INCLUDES := $(INCLUDES)`):

```
$ make BOARD=esp32-mh-et-live-minikit -C examples/hello-world/
make: Entering directory '/home/maribu/Repos/software/RIOT/examples/hello-world'
INCLUDES (cpu): file, recursive
INCLUDES (pkg): file, simple
INCLUDES = -isystem /usr/xtensa-esp32-elf/include/nano -I/home/maribu/Repos/software/RIOT/core/lib/include -I/home/maribu/Repos/software/RIOT/core/include -I/home/maribu/Repos/software/RIOT/drivers/include -I/home/maribu/Repos/software/RIOT/sys/include -I/home/maribu/Repos/software/RIOT/boards/esp32-mh-et-live-minikit/include -I/home/maribu/Repos/software/RIOT/cpu/esp32/include -I/home/maribu/Repos/software/RIOT/boards/common/esp32/include -I/home/maribu/Repos/software/RIOT/cpu/esp_common -I/home/maribu/Repos/software/RIOT/cpu/esp_common/include -I/home/maribu/Repos/software/RIOT/cpu/esp_common/include/freertos -I/home/maribu/Repos/software/RIOT/cpu/esp_common/vendor/ -I/home/maribu/Repos/software/RIOT/cpu/esp_common/vendor/esp -I/home/maribu/Repos/software/RIOT/cpu/esp32/esp-idf/include -I/home/maribu/Repos/software/RIOT/cpu/esp32/esp-idf/include/log -I/home/maribu/Repos/software/RIOT/cpu/esp32/vendor/include -I/components -I/components/bootloader_support/include -I/components/driver/include -I/components/esp_common/include -I/components/esp_hw_support/include -I/components/esp_hw_support/include/soc -I/components/esp_rom/esp32 -I/components/esp_rom/include -I/components/esp_rom/include/esp32 -I/components/esp_system/include -I/components/esp_system/port/include -I/components/esp_timer/include -I/components/hal/esp32/include -I/components/hal/include -I/components/hal/platform_port/include -I/components/heap/include -I/components/log/include -I/components/newlib/platform_include -I/components/soc/include -I/components/soc/esp32/include -I/components/xtensa/include -I/components/xtensa/esp32/include -I/home/maribu/Repos/software/RIOT/sys/libc/include
Building application "hello-world" for "esp32-mh-et-live-minikit" with MCU "esp32".

"make" -C /home/maribu/Repos/software/RIOT/pkg/esp32_sdk/ 
cc1: error: /components: No such file or directory [-Werror=missing-include-dirs]
[...]
```

In Docker, it stays recursively:

```
$ make BOARD=esp32-mh-et-live-minikit -C examples/hello-world/ BUILD_IN_DOCKER=1
make: Entering directory '/home/maribu/Repos/software/RIOT/examples/hello-world'
INCLUDES (cpu): file, recursive
INCLUDES (pkg): file, simple
INCLUDES = -isystem /usr/xtensa-esp32-elf/include/nano -I/home/maribu/Repos/software/RIOT/core/lib/include -I/home/maribu/Repos/software/RIOT/core/include -I/home/maribu/Repos/software/RIOT/drivers/include -I/home/maribu/Repos/software/RIOT/sys/include -I/home/maribu/Repos/software/RIOT/boards/esp32-mh-et-live-minikit/include -I/home/maribu/Repos/software/RIOT/cpu/esp32/include -I/home/maribu/Repos/software/RIOT/boards/common/esp32/include -I/home/maribu/Repos/software/RIOT/cpu/esp_common -I/home/maribu/Repos/software/RIOT/cpu/esp_common/include -I/home/maribu/Repos/software/RIOT/cpu/esp_common/include/freertos -I/home/maribu/Repos/software/RIOT/cpu/esp_common/vendor/ -I/home/maribu/Repos/software/RIOT/cpu/esp_common/vendor/esp -I/home/maribu/Repos/software/RIOT/cpu/esp32/esp-idf/include -I/home/maribu/Repos/software/RIOT/cpu/esp32/esp-idf/include/log -I/home/maribu/Repos/software/RIOT/cpu/esp32/vendor/include -I/components -I/components/bootloader_support/include -I/components/driver/include -I/components/esp_common/include -I/components/esp_hw_support/include -I/components/esp_hw_support/include/soc -I/components/esp_rom/esp32 -I/components/esp_rom/include -I/components/esp_rom/include/esp32 -I/components/esp_system/include -I/components/esp_system/port/include -I/components/esp_timer/include -I/components/hal/esp32/include -I/components/hal/include -I/components/hal/platform_port/include -I/components/heap/include -I/components/log/include -I/components/newlib/platform_include -I/components/soc/include -I/components/soc/esp32/include -I/components/xtensa/include -I/components/xtensa/esp32/include -I/home/maribu/Repos/software/RIOT/sys/libc/include
printf: ‘’: Invalid argument
Launching build container using image "docker.io/riot/riotbuild:latest".
podman run --rm --tty --userns keep-id -v '/etc/zoneinfo/Europe/Berlin:/etc/localtime:ro' -v '/home/maribu/Repos/software/RIOT:/data/riotbuild/riotbase:delegated' -v '/home/maribu/.cargo/registry:/data/riotbuild/.cargo/registry:delegated' -v '/home/maribu/.cargo/git:/data/riotbuild/.cargo/git:delegated' -e 'RIOTBASE=/data/riotbuild/riotbase' -e 'CCACHE_BASEDIR=/data/riotbuild/riotbase' -e 'BUILD_DIR=/data/riotbuild/riotbase/build' -e 'RIOTPROJECT=/data/riotbuild/riotbase' -e 'RIOTCPU=/data/riotbuild/riotbase/cpu' -e 'RIOTBOARD=/data/riotbuild/riotbase/boards' -e 'RIOTMAKE=/data/riotbuild/riotbase/makefiles'      -e 'BOARD=esp32-mh-et-live-minikit' -e 'DISABLE_MODULE=' -e 'DEFAULT_MODULE=' -e 'FEATURES_REQUIRED=' -e 'FEATURES_BLACKLIST=' -e 'FEATURES_OPTIONAL=' -e 'USEMODULE=' -e 'USEPKG='  -w '/data/riotbuild/riotbase/examples/hello-world/' 'docker.io/riot/riotbuild:latest' make 'BOARD=esp32-mh-et-live-minikit'    
INCLUDES (cpu): file, recursive
INCLUDES (pkg): file, recursive
INCLUDES = -I/data/riotbuild/riotbase/core/lib/include -I/data/riotbuild/riotbase/core/include -I/data/riotbuild/riotbase/drivers/include -I/data/riotbuild/riotbase/sys/include -I/data/riotbuild/riotbase/boards/esp32-mh-et-live-minikit/include -I/data/riotbuild/riotbase/cpu/esp32/include -I/data/riotbuild/riotbase/boards/common/esp32/include -I/data/riotbuild/riotbase/cpu/esp_common -I/data/riotbuild/riotbase/cpu/esp_common/include -I/data/riotbuild/riotbase/cpu/esp_common/include/freertos -I/data/riotbuild/riotbase/cpu/esp_common/vendor/ -I/data/riotbuild/riotbase/cpu/esp_common/vendor/esp -I/data/riotbuild/riotbase/cpu/esp32/esp-idf/include -I/data/riotbuild/riotbase/cpu/esp32/esp-idf/include/log -I/data/riotbuild/riotbase/cpu/esp32/vendor/include -I/data/riotbuild/riotbase/build/pkg/esp32_sdk/components -I/data/riotbuild/riotbase/build/pkg/esp32_sdk/components/bootloader_support/include -I/data/riotbuild/riotbase/build/pkg/esp32_sdk/components/driver/include -I/data/riotbuild/riotbase/build/pkg/esp32_sdk/components/esp_common/include -I/data/riotbuild/riotbase/build/pkg/esp32_sdk/components/esp_hw_support/include -I/data/riotbuild/riotbase/build/pkg/esp32_sdk/components/esp_hw_support/include/soc -I/data/riotbuild/riotbase/build/pkg/esp32_sdk/components/esp_rom/esp32 -I/data/riotbuild/riotbase/build/pkg/esp32_sdk/components/esp_rom/include -I/data/riotbuild/riotbase/build/pkg/esp32_sdk/components/esp_rom/include/esp32 -I/data/riotbuild/riotbase/build/pkg/esp32_sdk/components/esp_system/include -I/data/riotbuild/riotbase/build/pkg/esp32_sdk/components/esp_system/port/include -I/data/riotbuild/riotbase/build/pkg/esp32_sdk/components/esp_timer/include -I/data/riotbuild/riotbase/build/pkg/esp32_sdk/components/hal/esp32/include -I/data/riotbuild/riotbase/build/pkg/esp32_sdk/components/hal/include -I/data/riotbuild/riotbase/build/pkg/esp32_sdk/components/hal/platform_port/include -I/data/riotbuild/riotbase/build/pkg/esp32_sdk/components/heap/include -I/data/riotbuild/riotbase/build/pkg/esp32_sdk/components/log/include -I/data/riotbuild/riotbase/build/pkg/esp32_sdk/components/newlib/platform_include -I/data/riotbuild/riotbase/build/pkg/esp32_sdk/components/soc/include -I/data/riotbuild/riotbase/build/pkg/esp32_sdk/components/soc/esp32/include -I/data/riotbuild/riotbase/build/pkg/esp32_sdk/components/xtensa/include -I/data/riotbuild/riotbase/build/pkg/esp32_sdk/components/xtensa/esp32/include -I/data/riotbuild/riotbase/sys/libc/include
Building application "hello-world" for "esp32-mh-et-live-minikit" with MCU "esp32".

"make" -C /data/riotbuild/riotbase/pkg/esp32_sdk/ 
[...]
```

maribu commented:
> @aabadie: I also dropped the `export` as requested.

This causes issues in building the ESP32 bootloader, which uses `$(ESP32_SDK_DIR)` but does never include any relevant `Makefile.include`. I added the `export` again. Dropping the `export` should IMO be done in a separate PR, as this is more involved.

aabadie commented:
> Dropping the `export` should IMO be done in a separate PR, as this is more involved.

I fully agree!

gschorcht commented:
> This causes issues in building the ESP32 bootloader, which uses `$(ESP32_SDK_DIR)` but does never include any relevant `Makefile.include`.

That was exactly the reason why it was exported. Using RIOT's make files didn't work for building the bootloader, so it defines its own rules. `$(ESP32_SDK_DIR)` was defined in `pkg/esp32_sdk/Makefile.include` because it is related to the package.

maribu commented:
> `$(ESP32_SDK_DIR)` was defined in `pkg/esp32_sdk/Makefile.include` because it is related to the package.

I agree that this makes sense. I now know why `BUILD_IN_DOCKER=1` works for me: The include paths are still broken prior at the host, but inside the docker container `INCLUDES` remains recursively expended.

The difference between the docker and my local toolchain is that my local toolchain provides a nano flavor of newlib, which results in this being triggered:

https://github.com/RIOT-OS/RIOT/blob/2476a3b3a766dac0894d7f2447515e8d5f81528d/makefiles/libc/newlib.mk#L92-L102

which converts `INCLUDES` from recursively expended to simply expended. In the end, this issue may trigger for the Espressif provided as well once the provide a nano flavor. 

IMO we should make sure that the build process does not rely on a variable expending recursively. Even [GNU Make Manual](https://www.gnu.org/software/make/manual/html_node/Flavors.html) agrees with me that recursively expended variables are a pain in the ass:

> To avoid all the problems and inconveniences of recursively expanded variables, there is another flavor: simply expanded variables.

If we agree that we want to address the issue, I only see two options:

1. Move the definition of `ESP32_SDK_DIR` (as done here)
2. Change the order of how Makefiles are processed

I would advise to the latter, as previous experience showed that changing the order of how Makefiles are processed is a pain in the ass: Not only does it require quite some work to get it in a state that it is merged upstream, it often results in bugs popping up weeks after being merged :/

gschorcht commented:
> The difference between the docker and my local toolchain is that my local toolchain provides a nano flavor of newlib, which results in this being triggered

I encountered exactly the same problem a few months ago when I switched from my self-compiled toolchain to the vendor version of the toolchain. The Espressif toolchain provides the newlib-nano version, but does not use a separate include directory for the newlib-nano version. That's why I provided PR #17553 (commit d6f86a9b7ad33d15fb4b891ff0fc866cfe50cdfc) which overrides `INCLUDES` only if the separate include directory for the nano version exists.

gschorcht commented:
> I would advise to the latter, as previous experience showed that changing the order of how Makefiles are processed is a pain in the ass: Not only does it require quite some work to get it in a state that it is merged upstream, it often results in bugs popping up weeks after being merged :/

I am a little confused, you prefer option 2 despite of those problems?

maribu commented:
There is an "against" missing, sorry :)

gschorcht commented:
> There is an "against" missing, sorry :)

But the `export` issue isn't solved by option 1 :worried:

The test output of this PR still shows compilation issues. Are they solved with PR #18719?

maribu commented:
> But the `export` issue isn't solved by option 1

Indeed. The `export` is still present to get the builds running again. Let's tackle one issue at a time :)

maribu commented:
thx :)
